zebra-state/src/in_memory.rs

```diff
@@ -0,0 +1,119 @@
+use super::{Request, Response};
+use futures::prelude::*;
+use std::{
+    collections::{BTreeMap, HashMap},
+    error::Error,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+#[derive(Default)]
+struct ZebraState {
+    by_hash: HashMap<BlockHeaderHash, Arc<Block>>,
+    by_height: BTreeMap<BlockHeight, Arc<Block>>,
+}
+
+impl ZebraState {
+    fn insert(&mut self, block: impl Into<Arc<Block>>) {
+        let block = block.into();
+        let hash = block.as_ref().into();
+        let height = block.coinbase_height().unwrap();
+
+        assert!(
+            self.by_hash.insert(hash, block.clone()).is_none(),
+            "blocks shouldn't have the same hash"
+        );
+        assert!(
+            self.by_height.insert(height, block).is_none(),
+            "blocks with the same height are currently unsupported"
+        );
```

Comment:
If we want to disallow chain forks, I think it might be better to rearrange both of these checks so that we first check whether a block at that height exists; if not, insert, OK, if so, check that the hash matches, if so, OK, otherwise, error.

This also would let us send errors back through the service rather than panicing internally.